lib/rucio/client/uploadclient.py

```diff
@@ -366,8 +381,8 @@ def _collect_and_validate_file_info(self, items):
                 logger.warning('Skipping file %s because no rse was given' % path)
                 continue
             if pfn:
-                if item.get('no_register'):
-                    logger.warning('Upload with given pfn implies that no_register is True')
+                if item.get('no_register') and item.get('deterministic'):
```

Comment:
`item` probably does not have a 'deterministic' key. I would suggest to do this check later when you have the rse_settings

Reply:
Easiest solution:
          placing that part of code directly to upload() method.
Correct solution:
        move resolving rse_settings  to _collect_and_validate_file_info and publishing it to items

Reply:
Another solution:
  calling self.rses.setdefault(rse, rsemgr.get_rse_info(rse)) in _collect_and_validate method and    put  only is_deterministic flag to item dict.

Reply:
self.rses.setdefault(rse, rsemgr.get_rse_info(rse)) should be perhpas actually the first thing that we call in upload()

Reply:
No do it in the upload function please. I had a look and I remember why I did things that way. You loop 3 times over the files:
1. check if input is correct
2. check if all RSEs are ready
3. upload
This was to reduce the server calls in case something fails. It should be fine to do it in the second loop after the blacklisting check